### PR TITLE
Refactor `createCell`

### PR DESCRIPTION
**Pull request by jtoar**
While the diff of this PR may be intimidating, this PR introduces no new features and instead just cleans up the `createCell` file, commenting as much as I could (let me know if that gets annoying but I feel like every decision should be explained) and getting rid of a few things that I think were there only for historical reasons (i.e. before react hooks).

The only new thing this PR intentionally introduces is more a concise component-listing in the dev tools. Notice that there's no intermediate `Query` component:

**Before**

image

**After**

image

### CODE REVIEW

File: packages/web/src/components/createCell.tsx [hunk 1]
```diff
@@ -28,79 +24,184 @@ export type CellProps<
     (GQLVariables extends { [key: string]: never } ? unknown : GQLVariables)
 >
 
-export type DataObject = { [key: string]: unknown }
+type ScrubbedQueryOperationResult = Omit<
+  QueryOperationResult,
+  'loading' | 'error' | 'data'
+> & {
+  updating: boolean
+}
```
Comment (jtoar) [inline packages/web/src/components/createCell.tsx:32]: Tried reduce some of the duplication in the typing but didn't know what to name it

File: packages/web/src/components/createCell.tsx [hunk 2]
```diff
@@ -111,103 +212,115 @@ const isDataEmptyArray = (data: DataObject) => {
...
+       * If there's no `error` and there's no `data` and we're not `loading`,
+       * something's wrong. Most likely with the cache.
+       */
+      console.warn(
+        `If you're using Apollo Client, check for its debug logs here in the console, which may help explain the error.`
+      )
+      throw new Error(
+        'Cannot render Cell: reached an unexpected state where the query succeeded but `data` is `null`. If this happened in Storybook, your query could be missing fields; otherwise this is most likely a GraphQL caching bug. Note that adding an `id` field to all the fields on your query may fix the issue.'
+      )
+    }
```
Comment (jtoar) [inline packages/web/src/components/createCell.tsx:323]: May be a discussion for another time, but is there any reason not to just render `<Failure />` here as well?

File: packages/web/src/components/createCell.tsx [hunk 3]
```diff
@@ -111,103 +212,115 @@ const isDataEmptyArray = (data: DataObject) => {
...
+   * }
+   * ```
+   *
+   * `PostCell` is actually this component.
+   */
+  function NamedCell(props: React.PropsWithChildren<CellProps>) {
+    /**
+     * Right now, Cells don't render `children`.
+     */
+    const { children: _, ...variables } = props
```
Comment (jtoar) [inline packages/web/src/components/createCell.tsx:272]: May be a discussion for another time, but doing something like this seems conceivable to me:
```js
<PostsCell>
  <AddPost>
</PostsCell>
```
The difficult part about this is, where do we render it? My gut reaction is in both `<Empty />` and `<Success />`.

File: packages/web/src/components/createCell.tsx (hunk 2 above)
Comment (dac09) [reply packages/web/src/components/createCell.tsx:323]: Hmmm... you're right. Failure would make sense!

File: packages/web/src/components/createCell.tsx (hunk 1 above)
Comment (dac09) [reply packages/web/src/components/createCell.tsx:32]: yeah the name is difficult, but to me it sounds like this is the QueryOperationResult without any of the properties we add... but for some reason we also include updating 🤣

File: packages/web/src/components/createCell.tsx (hunk 3 above)
Comment (dac09) [reply packages/web/src/components/createCell.tsx:272]: Hmmm.... interesting.... @callingmedic911 and I had to do some workarounds for this (I think) when we did the filters on the tape dashboard.

Not needed for this PR, but we should bring up for discussion for sure.

**Comment by dac09** [comment]
@jtoar this looks like a very positive change, I'm all for it ✌️. Wondering if you've considered breaking this large file into smaller parts given how verbose it looks with the comments?

Perhaps we could separate the render blocks into individual functions e.g.:
- `renderDataSuccess`
- `renderUnknownError`
etc.

Not sure if it'll help here, just an idea!

File: packages/web/src/components/createCell.tsx [hunk 4]
```diff
@@ -111,103 +212,115 @@ const isDataEmptyArray = (data: DataObject) => {
...
-            throw new Error(
-              'Cannot render Cell: reached an unexpected state where the query succeeded but `data` is `null`. If this happened in Storybook, your query could be missing fields; otherwise this is most likely a GraphQL caching bug. Note that adding an `id` field to all the fields on your query may fix the issue.'
-            )
-          }
-        }}
-      </Query>
+  /**
+   * A Cell declaratively renders components based on the lifecycle of a GraphQL query.
+   *
+   * If the return's confusing, conider that a Cell's usually used in a Page like so:
```
Comment (callingmedic911) [inline packages/web/src/components/createCell.tsx:256]:
```suggestion
   * If the return's confusing, consider that a Cell's usually used in a Page like so:
```

BTW, I couln't understand this specific line. 🤔 Just me, maybe? 😅

Comment (jtoar) [reply packages/web/src/components/createCell.tsx:256]: Nah I think you're right, I was a little too into it. The new react docs have a line "This is why you should not nest component function definitions." I thought for a second that we were doing that but this is different. Happy to get rid of this one!

File: packages/web/src/components/createCell.tsx (hunk 1 above)
Comment (jtoar) [reply packages/web/src/components/createCell.tsx]: I reverted the change for now—if I have a better idea I'll add something but I don't yet

**Comment by dac09** [comment]
Yeah I think this is duplicated, but probably makes more sense

File: packages/web/src/components/createCell.tsx (hunk 1 above)
Comment (dac09) [reply packages/web/src/components/createCell.tsx]: Yeah makes sense 👍. I know its "duplicated" but seems more logical this way

**Comment by jtoar** [comment]
> Wondering if you've considered breaking this large file into smaller parts given how verbose it looks with the comments?

@dac09 I've shortened the comments a bit but it still takes up a chunk of the file. I'd like to keep it all in one file if possible still.

> Perhaps we could separate the render blocks into individual functions e.g.:
>
> - `renderDataSuccess`
> - `renderUnknownError`

I like this idea and have another PR where I explored something like it a bit. The return was something like:

```js
return (
  <React.Fragment>
    {renderLoading && <Loading ... />}
    {renderFailure && <Failure ... />}
    {renderEmpty && <Empty ... />}
    {renderSuccess && <Success ... />}
  </React.Fragment>
)
```

Making all those booleans got complicated, and I had to take care that I derived them all from the same source of state so that two things never appeared on the screen at the same time.

TL;DR (which should've come first) I think it's worth exploring but perhaps after this one? I tried to keep this one to just organizational changes so we can make the next step easier.

**Review by dac09 — APPROVED** [review]
Was just a passing comment @jtoar, but thank you for humouring me re:verbosity! About splitting the file - yeah I totally see why it might be hard to figure out, and I'm not against having it in one file if it makes more sense!

I think this is a big improvement as it is!

**Review by peterp — APPROVED** [review]
